Compiling 525.x264 with --AASteensgaardAgnostic results in error

When compiling `ldecod_src/img_process.c`, the compilation throws the following error:

```
jlc -c -o ldecod_src/img_process.o -DSPEC -DNDEBUG -Ildecod_src/inc -Ix264_src -Ix264_src/extras -Ix264_src/common -DSPEC_AUTO_SUPPRESS_OPENMP -DSPEC_AUTO_BYTEORDER=0x12345678   -JAASteensgaardAgnostic --JlmOptPassStatistics=print-steensgaard-analysis --JlmOptPassStatistics=print-memory-node-provisioning --JlmOptPassStatistics=print-basicencoder-encoding --JlmOptPassStatistics=print-dne-stat              -DSPEC_LP64 -fcommon  ldecod_src/img_process.c
terminate called after throwing an instance of 'jlm::util::error'                                      
  what():  Number of memory states cannot be zero.                                                     
specmake: *** [/home/reissman/Documents/spec-cpu2017-jlm/benchspec/Makefile.defaults:347: ldecod_src/img_process.o] Aborted (core dumped) 
specmake: *** Waiting for unfinished jobs....
```

The issue has potentially the same cause as #334.

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.cpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.cpp
@@ -813,10 +813,22 @@ public:
   }
 
   void
+  StartPointsToFlagsPropagationStatistics(const LocationSet & disjointLocationSet) noexcept
+  {
+    NumDisjointSets_ = disjointLocationSet.NumDisjointSets();
+    NumLocations_ = disjointLocationSet.NumLocations();
+    PointsToFlagsPropagationTimer_.start();
+  }
+
+  void
+  StopPointsToFlagsPropagationStatistics() noexcept
+  {
+    PointsToFlagsPropagationTimer_.stop();
+  }
+
+  void
   StartPointsToGraphConstructionStatistics(const LocationSet & locationSet)
   {
-    NumDisjointSets_ = locationSet.NumDisjointSets();
-    NumLocations_ = locationSet.NumLocations();
     PointsToGraphConstructionTimer_.start();
   }
 
@@ -893,6 +905,9 @@ public:
         "#UnknownMemorySources:",
         NumUnknownMemorySources_,
         " ",
+        "PointsToFlagsPropagationTime[ns]:",
+        PointsToFlagsPropagationTimer_.ns(),
+        " ",
         "PointsToGraphConstructionTime[ns]:",
         PointsToGraphConstructionTimer_.ns(),
         " ",
@@ -924,6 +939,7 @@ private:
   size_t NumUnknownMemorySources_;
 
   util::timer AnalysisTimer_;
+  util::timer PointsToFlagsPropagationTimer_;
   util::timer PointsToGraphConstructionTimer_;
   util::timer UnknownMemoryNodeSourcesRedirectionTimer_;
 };
@@ -1714,6 +1730,11 @@ Steensgaard::Analyze(
   // std::cout << LocationSet_->ToDot() << std::flush;
   statistics->StopSteensgaardStatistics();
 
+  // Propagate points-to flags in disjoint location set
+  statistics->StartPointsToFlagsPropagationStatistics(*LocationSet_);
+  PropagatePointsToFlags();
+  statistics->StopPointsToFlagsPropagationStatistics();
+
   // Construct PointsTo graph
   statistics->StartPointsToGraphConstructionStatistics(*LocationSet_);
   auto pointsToGraph = ConstructPointsToGraph();
@@ -1731,6 +1752,37 @@ Steensgaard::Analyze(
   LocationSet_.reset();
 
   return pointsToGraph;
+}
+
+void
+Steensgaard::PropagatePointsToFlags()
+{
+  bool pointsToFlagsChanged;
+  do
+  {
+    pointsToFlagsChanged = false;
+
+    for (auto & set : *LocationSet_)
+    {
+      auto location = set.value();
+
+      // Nothing needs to be done if this set does not point to another set
+      if (!location->GetPointsTo())
+      {
+        continue;
+      }
+      auto & pointsToLocation = LocationSet_->GetRootLocation(*location->GetPointsTo());
+
+      auto locationFlags = location->GetPointsToFlags();
+      auto pointsToLocationFlags = pointsToLocation.GetPointsToFlags();
+      auto combinedFlags = locationFlags | pointsToLocationFlags;
+      if (pointsToLocationFlags != combinedFlags)
+      {
+        pointsToLocation.SetPointsToFlags(combinedFlags);
+        pointsToFlagsChanged = true;
+      }
+    }
+  } while (pointsToFlagsChanged);
 }
 
 PointsToGraph::MemoryNode &

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.cpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.cpp
@@ -1730,7 +1730,7 @@ Steensgaard::Analyze(
   // std::cout << LocationSet_->ToDot() << std::flush;
   statistics->StopSteensgaardStatistics();
 
-  // Propagate points-to flags in disjoint location set
+  // Propagate points-to flags in disjoint location set graph
   statistics->StartPointsToFlagsPropagationStatistics(*LocationSet_);
   PropagatePointsToFlags();
   statistics->StopPointsToFlagsPropagationStatistics();

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
@@ -145,7 +145,7 @@ private:
   AnalyzeExtractValue(const rvsdg::simple_node & node);
 
   /**
-   * \brief Propagates the points-to flags throughout the disjoint set location graph.
+   * Propagates the points-to flags throughout the disjoint set location graph.
    *
    * AnalyzeRvsdg builds a disjoint set location graph, where each set is annotated with points-to
    * flags. This method propagates these points-to flags throughout the graph as all of these flags

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
@@ -147,7 +147,7 @@ private:
   /**
    * Propagates the points-to flags throughout the disjoint set location graph.
    *
-   * AnalyzeRvsdg builds a disjoint set location graph, where each set is annotated with points-to
+   * AnalyzeRvsdg() builds a disjoint set location graph, where each set is annotated with points-to
    * flags. This method propagates these points-to flags throughout the graph as all of these flags
    * are contagious. An example would be a set A that is marked as PointsToEscapedMemory and points
    * to another set B. As set B would "originate" from a load operation from "set A", the

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
@@ -145,6 +145,18 @@ private:
   AnalyzeExtractValue(const rvsdg::simple_node & node);
 
   /**
+   * \brief Propagates the points-to flags throughout the disjoint set location graph.
+   *
+   * Analyze builds a disjoint set location graph, where each set is annotated with points-to flags.
+   * This method propagates these points-to flags throughout the graph as all of these flags are
+   * contagious. An example would be a set A that is marked as PointsToEscapedMemory and points to
+   * another set B. As set B would "originate" from a load operation from "set A", the consequence
+   * would be that set B must be marked as PointsToEscapedMemory as well.
+   */
+  void
+  PropagatePointsToFlags();
+
+  /**
    * Construct a points-to graph from the Steensgaard analysis result.
    *
    * @return A points-to graph.

--- a/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
+++ b/jlm/llvm/opt/alias-analyses/Steensgaard.hpp
@@ -147,11 +147,11 @@ private:
   /**
    * \brief Propagates the points-to flags throughout the disjoint set location graph.
    *
-   * Analyze builds a disjoint set location graph, where each set is annotated with points-to flags.
-   * This method propagates these points-to flags throughout the graph as all of these flags are
-   * contagious. An example would be a set A that is marked as PointsToEscapedMemory and points to
-   * another set B. As set B would "originate" from a load operation from "set A", the consequence
-   * would be that set B must be marked as PointsToEscapedMemory as well.
+   * AnalyzeRvsdg builds a disjoint set location graph, where each set is annotated with points-to
+   * flags. This method propagates these points-to flags throughout the graph as all of these flags
+   * are contagious. An example would be a set A that is marked as PointsToEscapedMemory and points
+   * to another set B. As set B would "originate" from a load operation from "set A", the
+   * consequence would be that set B must be marked as PointsToEscapedMemory as well.
    */
   void
   PropagatePointsToFlags();

--- a/tests/TestRvsdgs.cpp
+++ b/tests/TestRvsdgs.cpp
@@ -3022,6 +3022,108 @@ MemcpyTest::SetupRvsdg()
 }
 
 std::unique_ptr<jlm::llvm::RvsdgModule>
+MemcpyTest2::SetupRvsdg()
+{
+  using namespace jlm::llvm;
+
+  auto rvsdgModule = RvsdgModule::Create(jlm::util::filepath(""), "", "");
+  auto rvsdg = &rvsdgModule->Rvsdg();
+
+  auto nf = rvsdg->node_normal_form(typeid(jlm::rvsdg::operation));
+  nf->set_mutable(false);
+
+  PointerType pointerType;
+  arraytype arrayType(pointerType, 32);
+  auto structBDeclaration = rvsdg::rcddeclaration::create({ &arrayType });
+  auto structTypeB = StructType::Create("structTypeB", false, *structBDeclaration);
+
+  auto SetupFunctionG = [&]()
+  {
+    iostatetype iOStateType;
+    MemoryStateType memoryStateType;
+    loopstatetype loopStateType;
+    FunctionType functionType(
+        { &pointerType, &pointerType, &iOStateType, &memoryStateType, &loopStateType },
+        { &iOStateType, &memoryStateType, &loopStateType });
+
+    auto lambda = lambda::node::create(rvsdg->root(), functionType, "g", linkage::internal_linkage);
+    auto s1Argument = lambda->fctargument(0);
+    auto s2Argument = lambda->fctargument(1);
+    auto iOStateArgument = lambda->fctargument(2);
+    auto memoryStateArgument = lambda->fctargument(3);
+    auto loopStateArgument = lambda->fctargument(4);
+
+    auto cFalse = jlm::rvsdg::create_bitconstant(lambda->subregion(), 1, 0);
+    auto c0 = jlm::rvsdg::create_bitconstant(lambda->subregion(), 32, 0);
+    auto c128 = jlm::rvsdg::create_bitconstant(lambda->subregion(), 64, 128);
+
+    auto gepS21 = GetElementPtrOperation::Create(s2Argument, { c0, c0 }, *structTypeB, pointerType);
+    auto gepS22 = GetElementPtrOperation::Create(gepS21, { c0, c0 }, arrayType, pointerType);
+    auto ldS2 = LoadNode::Create(gepS22, { memoryStateArgument }, pointerType, 8);
+
+    auto gepS11 = GetElementPtrOperation::Create(s1Argument, { c0, c0 }, *structTypeB, pointerType);
+    auto gepS12 = GetElementPtrOperation::Create(gepS11, { c0, c0 }, arrayType, pointerType);
+    auto ldS1 = LoadNode::Create(gepS12, { ldS2[1] }, pointerType, 8);
+
+    auto memcpyResults = Memcpy::create(ldS2[0], ldS1[0], c128, cFalse, { ldS1[1] });
+
+    auto lambdaOutput = lambda->finalize({ iOStateArgument, memcpyResults[0], loopStateArgument });
+
+    return std::make_tuple(lambdaOutput, jlm::rvsdg::node_output::node(memcpyResults[0]));
+  };
+
+  auto SetupFunctionF = [&](lambda::output & functionF)
+  {
+    iostatetype iOStateType;
+    MemoryStateType memoryStateType;
+    loopstatetype loopStateType;
+    FunctionType functionType(
+        { &pointerType, &pointerType, &iOStateType, &memoryStateType, &loopStateType },
+        { &iOStateType, &memoryStateType, &loopStateType });
+
+    auto lambda = lambda::node::create(rvsdg->root(), functionType, "f", linkage::external_linkage);
+    auto s1Argument = lambda->fctargument(0);
+    auto s2Argument = lambda->fctargument(1);
+    auto iOStateArgument = lambda->fctargument(2);
+    auto memoryStateArgument = lambda->fctargument(3);
+    auto loopStateArgument = lambda->fctargument(4);
+
+    auto functionFArgument = lambda->add_ctxvar(&functionF);
+
+    auto c0 = jlm::rvsdg::create_bitconstant(lambda->subregion(), 32, 0);
+
+    auto gepS1 = GetElementPtrOperation::Create(s1Argument, { c0, c0 }, *structTypeB, pointerType);
+    auto ldS1 = LoadNode::Create(gepS1, { memoryStateArgument }, pointerType, 8);
+
+    auto gepS2 = GetElementPtrOperation::Create(s2Argument, { c0, c0 }, *structTypeB, pointerType);
+    auto ldS2 = LoadNode::Create(gepS2, { ldS1[1] }, pointerType, 8);
+
+    auto callResults = CallNode::Create(
+        functionFArgument,
+        functionF.node()->type(),
+        { ldS1[0], ldS2[0], iOStateArgument, ldS2[1], loopStateArgument });
+
+    auto lambdaOutput = lambda->finalize(callResults);
+
+    rvsdg->add_export(lambdaOutput, { PointerType(), "f" });
+
+    return std::make_tuple(
+        lambdaOutput,
+        util::AssertedCast<CallNode>(jlm::rvsdg::node_output::node(callResults[0])));
+  };
+
+  auto [lambdaG, memcpyNode] = SetupFunctionG();
+  auto [lambdaF, callG] = SetupFunctionF(*lambdaG);
+
+  this->LambdaF_ = lambdaF->node();
+  this->LambdaG_ = lambdaG->node();
+  this->CallG_ = callG;
+  this->Memcpy_ = memcpyNode;
+
+  return rvsdgModule;
+}
+
+std::unique_ptr<jlm::llvm::RvsdgModule>
 LinkedListTest::SetupRvsdg()
 {
   using namespace jlm::llvm;

--- a/tests/TestRvsdgs.hpp
+++ b/tests/TestRvsdgs.hpp
@@ -1795,6 +1795,79 @@ private:
   jlm::rvsdg::node * Memcpy_;
 };
 
+/**
+ * This class sets up an RVSDG representing the following code snippet:
+ *
+ * \code{.c}
+ *   #include <string.h>
+ *
+ *   typedef struct structB
+ *   {
+ *     int * array[32];
+ *   } structB;
+ *
+ *   typedef struct structA
+ *   {
+ *     int x;
+ *     structB * b;
+ *   } structA;
+ *
+ *   static void
+ *   g(structB * s1, structB * s2)
+ *   {
+ *     memcpy(*s2->array, *s1->array, sizeof(int) * 32);
+ *   }
+ *
+ *   void
+ *   f(structA * s1, structA * s2)
+ *   {
+ *     g(s1->b, s2->b);
+ *   }
+ * \endcode
+ */
+class MemcpyTest2 final : public RvsdgTest
+{
+public:
+  [[nodiscard]] const jlm::llvm::lambda::node &
+  LambdaF() const noexcept
+  {
+    JLM_ASSERT(LambdaF_ != nullptr);
+    return *LambdaF_;
+  }
+
+  [[nodiscard]] const jlm::llvm::lambda::node &
+  LambdaG() const noexcept
+  {
+    JLM_ASSERT(LambdaG_ != nullptr);
+    return *LambdaG_;
+  }
+
+  [[nodiscard]] const jlm::llvm::CallNode &
+  CallG() const noexcept
+  {
+    JLM_ASSERT(CallG_ != nullptr);
+    return *CallG_;
+  }
+
+  [[nodiscard]] const jlm::rvsdg::node &
+  Memcpy() const noexcept
+  {
+    JLM_ASSERT(Memcpy_ != nullptr);
+    return *Memcpy_;
+  }
+
+private:
+  std::unique_ptr<jlm::llvm::RvsdgModule>
+  SetupRvsdg() override;
+
+  jlm::llvm::lambda::node * LambdaF_ = {};
+  jlm::llvm::lambda::node * LambdaG_ = {};
+
+  jlm::llvm::CallNode * CallG_ = {};
+
+  jlm::rvsdg::node * Memcpy_ = {};
+};
+
 /** \brief LinkedListTest class
  *
  * This class sets up an RVSDG representing the following code snippet:

--- a/tests/jlm/llvm/opt/alias-analyses/TestSteensgaard.cpp
+++ b/tests/jlm/llvm/opt/alias-analyses/TestSteensgaard.cpp
@@ -316,13 +316,14 @@ TestConstantPointerNull()
     auto & lambda = pointsToGraph.GetLambdaNode(*test.lambda);
     auto & lambdaOut = pointsToGraph.GetRegisterNode(*test.lambda->output());
     auto & lambdaArg = pointsToGraph.GetRegisterNode(*test.lambda->fctargument(0));
+    auto & externalMemoryNode = pointsToGraph.GetExternalMemoryNode();
 
     auto & constantPointerNull =
         pointsToGraph.GetRegisterNode(*test.constantPointerNullNode->output(0));
 
     assertTargets(lambdaOut, { &lambda });
-    assertTargets(lambdaArg, { &lambda, &pointsToGraph.GetExternalMemoryNode() });
-    assertTargets(constantPointerNull, {});
+    assertTargets(lambdaArg, { &lambda, &externalMemoryNode });
+    assertTargets(constantPointerNull, { &lambda, &externalMemoryNode });
 
     jlm::util::HashSet<const jlm::llvm::aa::PointsToGraph::MemoryNode *> expectedEscapedMemoryNodes(
         { &lambda });
@@ -1087,6 +1088,43 @@ TestMemcpy()
 }
 
 static void
+TestMemcpy2()
+{
+  // Arrange
+  jlm::tests::MemcpyTest2 test;
+  // jlm::rvsdg::view(test.graph().root(), stdout);
+
+  // Act
+  auto pointsToGraph = RunSteensgaard(test.module());
+  std::cout << jlm::llvm::aa::PointsToGraph::ToDot(*pointsToGraph) << std::flush;
+
+  // Assert
+  assert(pointsToGraph->NumRegisterNodes() == 8);
+  assert(pointsToGraph->NumLambdaNodes() == 2);
+
+  auto & lambdaFNode = pointsToGraph->GetLambdaNode(test.LambdaF());
+  auto & lambdaFArgument0 = pointsToGraph->GetRegisterNode(*test.LambdaF().fctargument(0));
+  auto & lambdaFArgument1 = pointsToGraph->GetRegisterNode(*test.LambdaF().fctargument(1));
+
+  auto & lambdaGArgument0 = pointsToGraph->GetRegisterNode(*test.LambdaG().fctargument(0));
+  auto & lambdaGArgument1 = pointsToGraph->GetRegisterNode(*test.LambdaG().fctargument(1));
+
+  auto & memcpyOperand0 = pointsToGraph->GetRegisterNode(*test.Memcpy().input(0)->origin());
+  auto & memcpyOperand1 = pointsToGraph->GetRegisterNode(*test.Memcpy().input(1)->origin());
+
+  auto & externalMemoryNode = pointsToGraph->GetExternalMemoryNode();
+
+  assertTargets(lambdaFArgument0, { &lambdaFNode, &externalMemoryNode });
+  assertTargets(lambdaFArgument1, { &lambdaFNode, &externalMemoryNode });
+
+  assertTargets(lambdaGArgument0, { &lambdaFNode, &externalMemoryNode });
+  assertTargets(lambdaGArgument1, { &lambdaFNode, &externalMemoryNode });
+
+  assertTargets(memcpyOperand0, { &lambdaFNode, &externalMemoryNode });
+  assertTargets(memcpyOperand1, { &lambdaFNode, &externalMemoryNode });
+}
+
+static void
 TestLinkedList()
 {
   auto validatePointsToGraph = [](const jlm::llvm::aa::PointsToGraph & pointsToGraph,
@@ -1187,6 +1225,7 @@ TestSteensgaardAnalysis()
   TestEscapedMemory3();
 
   TestMemcpy();
+  TestMemcpy2();
 
   TestLinkedList();
 

